adal/src/androidTest/java/com/microsoft/aad/adal/AcquireTokenRequestTest.java

```diff
@@ -558,6 +599,433 @@ public void testVerifyBrokerRedirectUriInvalidSignature() throws PackageManager.
+     * The request should be posted twice with 500/503/504 error from server
+     */
+     public void testExtendedLifetimeRequest_retry_positive() throws PackageManager.NameNotFoundException,
+     NoSuchAlgorithmException, OperationCanceledException, IOException, AuthenticatorException,
+     InterruptedException {
+         //prepare cache
+         final Calendar expiredTime = new GregorianCalendar();
+         expiredTime.add(Calendar.MINUTE, -MINUS_MINUITE);
+         final Calendar extendedExpiresTime = new GregorianCalendar();
+         extendedExpiresTime.add(Calendar.MINUTE, 0);
+         final ITokenCacheStore cacheStore = getExtendedLifetimeTokenCache(expiredTime.getTime(), false, false, extendedExpiresTime.getTime());
+
+         //prepare context
+         final AccountManager mockedAccountManager = getMockedAccountManager();
+         mockAccountManagerGetAccountBehavior(mockedAccountManager);
+         final FileMockContext mockContext = createMockContext();
+         mockContext.setMockedAccountManager(mockedAccountManager);
+         final AuthenticationContext authContext = new AuthenticationContext(mockContext,
+                 VALID_AUTHORITY, false, cacheStore);
+         authContext.setExtendedLifetimeEnabled(true);
+
+         //mock http response
+         final HttpURLConnection mockedConnection = Mockito.mock(HttpURLConnection.class);
+         HttpUrlConnectionFactory.mockedConnection = mockedConnection;
+         Util.prepareMockedUrlConnection(mockedConnection);
+         Mockito.when(mockedConnection.getOutputStream()).thenReturn(Mockito.mock(OutputStream.class));
+         Mockito.when(mockedConnection.getInputStream()).thenReturn(Util.createInputStream(Util.getSuccessTokenResponse(false, false)),
+                 Util.createInputStream(Util.getSuccessTokenResponse(true, true)));
+         Mockito.when(mockedConnection.getResponseCode()).thenReturn(HttpURLConnection.HTTP_GATEWAY_TIMEOUT);
+
+         try {
+             authContext.acquireTokenSilentSync("resource", "clientid", TEST_USERID);
+             fail("Not expected exception.");
+         } catch (final AuthenticationException exception) {
+             //verify the retry
+             verify(mockedConnection, times(2)).getInputStream();
+             assertTrue(exception.getCode() == ADALError.AUTH_FAILED_NO_TOKEN);
+             assertTrue(exception.getCause() instanceof AuthenticationException);
+             assertTrue(((AuthenticationException)exception.getCause()).getCode() == ADALError.SERVER_ERROR_FOR_RETRY);
+         } finally {
+             cacheStore.removeAll();
+         }
+     }
+
+    /**
+     * Test for the retry when encountering with non-500/503/504 server error, eg. 502 HTTP_BAD_GATEWAY
+     * The request should not retry if we receive other error rather than 500/503/504 from server
+     */
+    public void testExtendedLifetimeRequest_retry_negative() throws PackageManager.NameNotFoundException,
```

Comment:
It is used to testing the retry logic we do not address the testing of return here.